src/python/pants/task/task.py

```diff
@@ -96,7 +97,7 @@ def _compute_stable_name(cls):
   @classmethod
   def subsystem_dependencies(cls):
     return (super(TaskBase, cls).subsystem_dependencies() +
-            (CacheSetup.scoped(cls), BuildInvalidator.Factory, SourceRootConfig))
+            (CacheSetup.scoped(cls), TargetFilter.scoped(cls), BuildInvalidator.Factory, SourceRootConfig))
```

Comment:
This should probably be conditional on whether a `Task` actually opts-in to Target filtering: as `get_targets` notes, very few tasks actually do, and for many `Tasks`, target filtering doesn't make sense. Having this on "by default" like this will make it show up in the help and documentation for a whole bunch of `Tasks` for which it will have no effect.

As with @wisechengyi's initial patch, I would suggest a classmethod/classproperty of `Task` that subclasses could override to enable filtering (with filtering disabled by default), and then conditionally requesting the subsystem here if that property was True.

Reply:
Good point! Yes, this will be a necessary follow-up. 

But does this subsystem-based design address your v2 engine concerns?

Reply:
It does, thanks! This is very easy to transition to v2.

Reply:
👍 Addressing this in https://github.com/pantsbuild/pants/pull/7283